### PR TITLE
Week-1

### PR DESCRIPTION
**Pull request by archanarw**

### CODE REVIEW

**Comment by mschauer** [comment]
Very nice, I could use some words taking me through this.  1.) The `mechanism`s inherited from Omega, can I think of them as Markov kernels?  2.) Differentiability with respect to which variables are you talking about?

**Comment by archanarw** [comment]
1. `mechanism` is a function I wrote to return the SCM at each node in `CausalModel`. In "Causal Inference in Statistics: A Primer", the authors have called the functions mechanisms, so I used the same name.
2. Differentiability with respect to exogenous and endogenous variables in the model.

File: src/causalmodel.jl [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    indegree, outdegree, eltype
+
+# :name -> variable_name and :func -> function or distribution
+const Variable = @NamedTuple{name::Symbol, func}
+
+"Causal model with `dag` representing the model and `scm` holds the name of the variables and the SCM"
+struct CausalModel{T} <: AbstractGraph{T}
+    dag::SimpleDiGraph{T}  # Representing the DAG of the causal model
+    scm::Dict{T,Variable}  # Name and function (or distribution) of each variable
+end
```
Comment (mschauer) [inline src/causalmodel.jl:28]: I like this approach, this is even compatible with Soss' conditional models in the full `causalmodel.jl` file

File: test/interventions.jl [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    intervention_ = DifferentiableIntervention(x, β, i.l)
+    output = apply_intervention(g, intervention_)
+    loss_ = sum(output)
+    return loss_
+end
+@test typeof(ForwardDiff.gradient(loss, vcat(i.x, i.β))) == Array{Float64,1}
+
+i = Intervention(:SAT, 1560)
+m = apply_intervention(g, i)
+@test typeof(m) == CausalModel{Int64}
```
Comment (vollmersj) [inline test/interventions.jl:35]: Why shouuld this type be parametrized by Int64

Comment (archanarw) [reply test/interventions.jl:35]: `CausalModel` is a subtype of `AbstractGraph`, and that is parameterized in a similar way.

**Comment by zenna** [comment]
Some high level comments.

It shouldn't be the case that you have Omega random variables as the nodes of your causal graph, instead you should implement the interfaces of Omega.  These are

1. Application of a model or random variable to \omega
`m(\omega)` where M :: CausalModel

```Julia
sing CounterfactualFairness, Test
using Omega, Distributions

# U₁ = 1 ~ Normal(27, 5)
# U₂ = 2 ~ Normal(2, 4)
# U₃ = 3 ~ Normal(1, 2)

# X = U₁
# Ya(ω) = 4 * X(ω)
# Yb(ω) = Ya(ω) + U₂(ω)
# Z(ω) = X(ω) / U₃
g = CausalModel()
U₁ = add_exo_variable!(g, :U₁, 1 ~ Normal(27, 5))
U₂ = add_exo_variable!(g, :U₂, 2 ~ Normal(27, 5))
U₃ = add_exo_variable!(g, :U₃, 3 ~ Normal(27, 5))
X = add_endo_variable!(g, :X, identity, U₁)
Ya = add_endo_variable!(g, :Ya, *, 4, X)
Yb = add_endo_variable!(g, :Yb, +, Ya, U₂)
Z = add_endo_variable!(g, :Yb, /, X, U₃)

@test typeof(X) == CausalVar
output = apply_context(g, (U₁ = 1.23, U₂ = 15, U₃ = 1.451))

ω = defω()
output2 = g(ω)
@test typeof(output) == typeof(output2)

@test typeof(X(ω)) == Float64
@test X(ω) == output2.X 

g = add_vertex(g, (:Temp, identity))
g = add_vertex(g, (:IceCreamSales, Y))
g = add_vertex(g, (:Crime, Z))

Z_intervene = intervene(Z. X => 12.30)
Z_intervene(ω)
@test Z_intervene != Z(ω)
```

**Pull request merged**